Capture modularisation/cleanup

johnnovak opened the pull request:
Hacking'n'slashing the entangled spaghetti mess responsible for capturing stuff into more manageable pieces.
No functional changes, just reorganising stuff & cleanup. AFAIK I haven't broken anything, only upset the linters a bit... (to be fixed).

This all lived in `hardware.cpp` previously in a big cosy ball of mud 😩 Sure, that's a good place... we're all dealing with "hardware" after all... computers are hardware too, and so on 🤦🏻 🥶 

This is in preparation for the screenshot improvements.

kcgen commented:
> This all lived in hardware.cpp previously in a big cosy ball of mud weary Sure, that's a good place... we're all dealing with "hardware" after all... computers are hardware too, and so on 🤦🏻 :cold_face: 

:rofl:

ThomasEricB commented:
Nice work @johnnovak ! You really showed `hardware.cpp` who is boss!

johnnovak commented:
> You really showed `hardware.cpp` who is boss!

It had it coming. 😎

johnnovak commented:
So, for the record, this builds fine on macOS, even after a full clean & rebuild. I'll screw around on Windows today or tomorrow to eradicate those build failures—hopefully that will get rid of most of the Linux failures too.

dreamer commented on include/hardware.h:
```diff
@@ -1,5 +1,5 @@
 /*
- *  Copyright (C) 2002-2021  The DOSBox Team
+ *  Copyright (C) 2002-2023  The DOSBox Team
```
We are keeping "DOSBox Team" copyrights intact; in cases like this one, add another copyright line for "DOSBox Staging Team" (grep through the code will give plenty of examples).

johnnovak replied:
Yeah okay. TBH I did not even read it, just updated the date 😄

johnnovak replied:
Done in all files.

dreamer commented on src/capture/capture.cpp:
```diff
@@ -0,0 +1,268 @@
+	}
+	close_directory(dir);
+	char file_name[CROSS_LEN];
+	sprintf(file_name,
```
Let's update it to `safe_sprintf` while we're at it.

johnnovak replied:
Sure, good idea.

johnnovak replied:
Done.

dreamer commented on src/capture/capture.cpp:
```diff
@@ -0,0 +1,268 @@
+
+	FILE* handle = fopen(file_name.c_str(), "wb");
+	if (handle) {
+		LOG_MSG("Capturing %s to %s", type, file_name.c_str());
```
Let's surround paths we print to log with quotes: `'%s'`.

johnnovak replied:
Good idea.

johnnovak replied:
Done.

johnnovak commented:
Moving this back to draft as I'm going to clean it up more substantially in addition to the fairly mechanical splitting/refactoring/etc. I've already done.

The end goal is to make all this easily modifiable for future enhancements... because currently it isn't 😩

johnnovak commented:
Okay, this is ready for final review @kcgen & @dreamer. I've regressions tested everything and all capture functionality works 100% fine.

I think the whole thing should be condensed into a single commit; the commit history doesn't add much value as files have been split, moved around, and refactored in a major way. Plus I did break a few things in the first phase because the state handling was still a mess... This has been rectified in my last big refactor commit. Now you can mostly follow what's going on by reading the code, finally 😅

johnnovak commented:
Someone give me a tick please? @kcgen @dreamer @FeralChild64 Feel free to test, I could not spot any regressions/problems after the refactor.

kcgen commented:
The "Move parts of generate rendered screenshot to capture_image" commit fails to compile (bad for bisecting):

``
In file included from ../../include/sdlmain.h:33,
                 from ../../src/capture/capture_image.cpp:24:
../../src/capture/../libs/ppscale/ppscale.h: In constructor ‘PPScale::PPScale(int, int, double, bool, int, int)’:
../../src/capture/../libs/ppscale/ppscale.h:51:17: error: ‘assert’ was not declared in this scope
   51 |                 assert(source_w > 0);
      |                 ^~~~~~
```

kcgen reviewed (commented):
Tests are running well - I tested wav, avi, and both screenshots all individually and then all combinations:
wav
avi
screenshots
wav + avi
wav + screenshots
wav + avi + screenshots
avi + screenshots

So if this is cleanup only without functional changes intended, then I think it's OK.  Just need all commits building and it'll be set!

johnnovak commented:
Thanks for testing it @kcgen. I've squashed the offending commits, I'm pretty sure I was able to compile every commit on macOS, but they I was relying on incremental commits.

kcgen commented:
> Thanks for testing it @kcgen. I've squashed the offending commits, I'm pretty sure I was able to compile every commit on macOS, but then I was relying on incremental compilation.

:+1: , your good commit messages indicate unbuildable commits still exist in the latest push:

![2023-05-09_08-54](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/97c56efd-c5af-43ca-bb6d-c1c1a4847076)

Checking prior-to-fixed macOS commit -- indeed, it doesn't build:

![2023-05-09_08-58](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/f485a71c-6ea9-4174-b582-cfa703be11f3)

kcgen reviewed (approved):
approving - feel free to merge @johnnovak once those last compile commits are handled :rocket:

kcgen commented:
> I think the whole thing should be condensed into a single commit

Sorry I missed that earlier! Given it all checks out, I'll go-ahead and squash and merge.

johnnovak commented:
> > I think the whole thing should be condensed into a single commit
> 
> Sorry I missed that earlier! Given it all checks out, I'll go-ahead and squash and merge.

Yeah, by the final commit everything ended up being substantially refactored anyway, so hence my suggestion to just squash the whole thing. It would be possible to do such big refactors and keep all builds green on the different plstforms, but that's much more time consuming... Thanks for merging.
